### ISSUE TITLE
NaN values in reward

### ISSUE THREAD

**Issue by darnox** [Unity-Technologies/ml-agents#117]
Hi,

I'm training my model but at some point, a strange thing happens. The reward returned by PPO sometimes has 'nan' value. It occurs in different places (not necessarily on specific epoch) and after a few steps, it can get back to correct float values and then once again return 'nan' in a future. It doesn't apply only to _Mean Reward_, the _cumulative_reward_ shown in tensorboard is broken at this epochs as well. I'm attaching part of an output from training process. Maybe someone knows why it's happening.

Thanks 

> Mean Reward: -82.99281905388821
> Saved Model
> Mean Reward: -48.54445689334988
> Saved Model
> Mean Reward: -37.32348710034999
> Saved Model
> D:\ProgramData\Miniconda3\envs\unity\lib\site-packages\numpy\core\fromnumeric.py:2909: RuntimeWarning: Mean of empty slice.
>   out=out, **kwargs)
> D:\ProgramData\Miniconda3\envs\unity\lib\site-packages\numpy\core\_methods.py:80: RuntimeWarning: invalid value encountered in double_scalars
>   ret = ret.dtype.type(ret / rcount)
> Mean Reward: nan
> Saved Model
> Mean Reward: nan
> Saved Model
> Mean Reward: nan
> Saved Model

**Comment by vincentpierre** [comment]
Hi, it is possible that you ask for a report on the mean reward but no data point was collected yet, hence a division by 0. Can you try increasing `summary-freq` and `save-freq` in the ppo arguments and see if it solves your problem ?

**Comment by surferau** [comment]
I had nan error when all episodes were consisting exactly from one step (every AgentStep call had "done"  property  set to true)

**Comment by MhdTlb** [comment]
Can I work on this?

**Comment by marwie** [comment]
I seem not to be able to get rid of this error as well and I'm not sure what is causing it.
My agent is quite simple: its a ball with a rigidbody and applies a impulse force on x and z vectors (continuous).
The state contains the start position (for each vector3 I only use xz), the current position, the target position, the obstacle positions I want the agent to avoid as well as the current velocity.

the step method looks like this at the moment

```
 public override void AgentStep(float[] action)
    {
        var currentDistance = GetDistance();
        if (currentDistance < .5f)
        {
            reward = 1;
            done = true;
            return;
        }
        else if(rigidBody.velocity.magnitude > 30)
        {
            reward = -1;
            done = true;
            return;
        }

        float forceX = action[0];
        float forceZ = action[1];
        var f = new Vector3(forceX, 0, forceZ);
        rigidBody.AddForce(f, ForceMode.Impulse);

        var currentObstaclesDistance = GetClosestObstaclesDistance();

        float threshold = 0.1f;
        bool closerToTarget = currentDistance + threshold < previousTargetDistance;

        if (GetClosestObstaclesDistance() < 0.5f)
        {
            reward = -.1f;
        }
        else if (closerToTarget)
        {
            previousTargetDistance = currentDistance;
            reward = 0.1f;
        }
    }
```

the error I randomly keep getting is:

```
C:\Users\user\Anaconda3\lib\site-packages\numpy\core\fromnumeric.py:2909: RuntimeWarning: Mean of empty slice.
  out=out, **kwargs)
C:\Users\user\Anaconda3\lib\site-packages\numpy\core\_methods.py:80: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
```

on reset the obstacles, target and agent position are randomly regenerated making sure they not too close to each other.

Maybe I miss something totally obvious :)

**Comment by marwie** [comment]
ok I think my mistake was probably having MaxStep set to 0?

**Comment by GuSoder** [comment]
I also have this problem, I encountered it when changing my action state vector, which initially seemed weird. But then I noticed I had changed the probability of my scripts setting academy.done to true, so I guess the episodes got to small to capture this. It seems I get this problem less often (less nan results) by setting max time steps in the academy object, and setting the summary frequency to five times that number, I still get some nan mean results though, and I think I always set a reward for each agent.

**Comment by raeldor** [comment]
I'm getting the same problem here.  Does anyone know what causes this?  I even tried assigning a static reward each agent step but still it comes back as NaN even though I have MaxStep set.  I don't think the agent is getting done state because I have reset on done set and reset causes a visible reset that I'm not seeing.  Really stumped on this one.

**Comment by raeldor** [comment]
Well, it looks like setting Max Steps on the agent isn't doing anything.  Only setting Max Steps on the academy seems to force the agent to done. :(

**Comment by raeldor** [comment]
So, this is still messing me up.  I've been stepping through the code in Heuristic mode and noticed that my reward is being reset by...
        foreach (Brain brain in brains)
        {
            brain.ResetDoneAndReward();
        }

In the Academy Step() method before the reward is being added by the call to SetCumulativeReward() inside CollectStates() which is called from DecideAction().  Tracing this back to RunMdp() it seems that step() is called, which resets the reward BEFORE the call to DecideAction() which gets the states and adds the reward...

                else
                {
                    Step();
                }
            }

            DecideAction();

Why on earth would it be resetting the reward before collecting the states and adding the reward to the cumulative reward?!

**Comment by Lakrix** [comment]
I get the nan values when I increase the Time Scale together with decreasing the Quality Level, Time Scale = 100 and Quality Level = 0. If I used Quality Level as 5 and Time Scale as 1 everything works fine.

**Comment by vincentpierre** [comment]
Hi, we released version 0.3 of ml-agents, please let us know if you are still facing this issue.
